dxr/plugins/rust/menu.py

```diff
@@ -34,106 +35,187 @@ def truncate_value(value, typ=""):
+
+    """
+    def __init__(self, tree, path, row):
+        super(_JumpToTarget, self).__init__(tree)
+        self.path = path
+        self.row = row
+
+    def es(self):
+        return self.path, self.row
+
+    @classmethod
+    def from_decl(cls, tree, decl):
+        """Return an instance made from a declaration mapping.
+
+        If the incoming declaration doesn't warrant the creation of a menu,
+        return None.
+
+        """
+        path = decl['file_name']
+        if path:
+            return cls(tree, path, decl['file_line'])
+        else:
+            warn("Can't add jump to empty path.")  # Can this happen?
+
+    def menu_items(self):
+        yield {'html': 'Jump to %s' % self.target_name,
+               'title': "Jump to %s in '%s'" % (self.target_name,
+                                                os.path.basename(self.path)),
+               'href': url_for(BROWSE, tree=self.tree.name, path=self.path, _anchor=self.row),
+               'icon': 'jump'}
+
+
+class JumpToTraitMethodMenuMaker(_JumpToTarget, _RustPluginAttr):
+    target_name = 'trait method'
+
+
+class JumpToDefinitionMenuMaker(_JumpToTarget, _RustPluginAttr):
+    target_name = 'definition'
+
+
+class JumpToModuleDefinitionMenuMaker(_JumpToTarget, _RustPluginAttr):
+    target_name = 'module defintion'
+
+
+class JumpToModuleDeclarationMenuMaker(_JumpToTarget, _RustPluginAttr):
+    target_name = 'module declaration'
 
-def function_menu(tree, datum, tree_config):
-    menu = function_menu_generic(tree, datum, tree_config)
 
+class JumpToAliasDefinitionMenuMaker(_JumpToTarget, _RustPluginAttr):
+    target_name = 'alias defintion'
+
+
+class JumpToCrateMenuMaker(_JumpToTarget, _RustPluginAttr):
+    target_name = 'crate'
+
+
+class JumpToTypeDeclarationMenuMaker(_JumpToTarget, _RustPluginAttr):
+    target_name = 'type declaration'
+
+
+class JumpToVariableDeclarationMenuMaker(_JumpToTarget, _RustPluginAttr):
+    target_name = 'variable declaration'
+
+
+class JumpToFunctionDeclarationMenuMaker(_JumpToTarget, _RustPluginAttr):
+    target_name = 'function declaration'
+
+
+class TraitImplMenuMaker(SingleDatumMenuMaker):
+    def menu_items(self):
+        qualname, count = self.data
+        yield {'html': "Find implementations (%d)" % self.count,
+               'title': "Find implementations of this trait method",
+               'href': search_url(self.tree, "+fn-impls:%s" % quote(self.qualname)),
+               'icon': 'method'}
+
+
+def function_menu(tree, datum, tree_config):
+    makers = []
     if 'declid' in datum and datum['declid'] in tree.data.functions:
-        # it's an implementation, find the decl
+        # It's an implementation; find the decl:
         decl = tree.data.functions[datum['declid']]
-        add_jump_definition(tree, tree_config, menu, decl, "Jump to trait method")
+        makers.append(JumpToTraitMethodMenuMaker.from_decl(tree_config, decl))
     else:
         # it's a decl, find implementations
```

Comment:
For symmetry, may as well also make this "It's a declaration; find implementations:"